### DISCUSSION TITLE
Child Node Slots

### DISCUSSION THREAD

**Discussion by TitanNano** [godotengine/godot-proposals#4374] [Scripting]
## What?
I think it would be quite usefull we we could replace:

```gdscript
onready var child_node = $ChildNode
```

With something like:
```gdscript
# we can define a slot name for the editor to display as well as the desired node type
slot("ChildSlot", Spatial) var child_node: NodeSlot 

# a slot also can hold multiple nodes but the default flag would be ALLOW_SINGLE
slot("ParticleSlot", Particles, ALLOW_MULTIPLE) var particles: NodeSlot 
```

The editor then can display slots among the child nodes, but perhaps a slot will be replaced once a child is assigned. The exact UX has to be designed first.

<img alt="slot example" src="https://user-images.githubusercontent.com/4500858/162619655-9b368191-f95e-486d-9910-5754cea10007.png" width="296">

## Why?
This would allow us to decouple the script from the node tree, by removing the need for a known node name. It would also allow for "abstract" / inherited scenes to declare which child nodes have to be added to make it functional.

## Additional
There would also be a need for the editor to support slots by displaying them in the scene tree and let us assign child nodes to these slots.

**Comment by Zireael07** [comment]
How is that different from assigning Node Paths? How do I assign to a slot?

**Reply by TitanNano** [reply]
Assigning node paths works but is a bit cumbersome as we have to both add the child node and assign it's node path. With slots the script can declare it's required child nodes and the user can just assign the matching nodes accordingly. It could work somewhat like this:
<img width="296" alt="image" src="https://user-images.githubusercontent.com/4500858/162619655-9b368191-f95e-486d-9910-5754cea10007.png">
The editor displays a slot and we can assign children to it.

**Comment by Mickeon** [comment]
This sounds like a more complicated and vastly less integrated version of https://github.com/godotengine/godot-proposals/issues/1048 or https://github.com/godotengine/godot-proposals/issues/934 .

Regarding not having to worry about a node name, it should be something the editor manages for the user, as it already currently does for exported NodePaths.

**Reply by TitanNano** [reply]
I guess the main advantage of slots would be that you'd only have to assign a child node to the right slot and not to the exported property as well. See https://github.com/godotengine/godot-proposals/discussions/4374#discussioncomment-2539379 for an example.
